text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+a proper TLS server::
+
+    --tls-certfile
+    --tls-keyfile
+    --tls-private-key-password
+    --tls-protocol-min
+    --tls-protocol-max
+    --tls-ciphers
+
+The end user is not usually supposed to interact with this API directly, so we
+need to bring this to a higher user-level interface with convenience in mind.
+We shall discuss about that later in this RFC.
+
+The EdgeDB TLS server will by default advertise all protocols supported by the
+server using the ALPN Extension during TLS handshakes. However the way we
+advertise them is still pending for discussion. Protocols we have now:
+
+* The EdgeDB binary protocol
+
+* HTTP protocol, on top of which we have:
+
+  * System API starting with ``/server```
+
+  * Extensions that may differ from databases, starting with ``/db``:
+
+    * EdgeQL over HTTP
+
+    * GraphQL over HTTP
+
+    * "Notebook" - parameterless EdgeQL in HTTP-posted JSON
+
+All EdgeDB client bindings should use TLS to talk to an EdgeDB server.
+If possible, the client should also join the ALPN negotiation to choose its
+preferred protocol for each connection. The ALPN support in target languages:
+
+* Python [4]_: ``set_alpn_protocols()`` and ``selected_alpn_protocol()``
+* Go [5]_: ``SupportedProtos`` and ``NegotiatedProtocol``
+* Node.js [6]_: ``ALPNProtocols`` and ``alpnProtocol``
+
+If a specific protocol was successfully negotiated, that protocol will be used
+as the exclusive one throughout the lifetime of the connection. Or else if the
+ALPN agreement was not made, depending on the discussion below we will fallback
+to either the EdgeDB binary protocol or the current HTTP protocol.
+
+
+Development of EdgeDB
+---------------------
+
+As mentioned in the Overview, the ``edb server`` CLI will take extra parameters
+for TLS. We will use the Python builtin TLS support to handle the certificates
+and ALPN, and use the TLS transport implementation in uvloop for the network.
+
+The server-side ``ssl.SSLContext`` [4]_ will be initialized with the default
+``protocol=ssl.PROTOCOL_TLS``, leaving the control of accepted TLS protocol
+versions to ``SSLContext.minimum_version`` and ``SSLContext.maximum_version``,
+which in turn are managed by:
+
+* ``--tls-protocol-min``: One of ``1.0`` (default), ``1.1`` and ``1.2``
+* ``--tls-protocol-max``: One of ``1.0``, ``1.1``, ``1.2`` (default) and ``1.3``
+
+SSLv2 and SSLv3 are not supported. TLSv1.3 support is included (question:
+shall/can we deliver with OpenSSL >= 1.1.1 consistently for all platforms?).
+
+``--tls-certfile``, ``--tls-keyfile`` and ``--tls-private-key-password`` are
+directly the parameters of Python ``ssl.SSLContext.load_cert_chain()``,
+except that the password cannot be a function. Instead, the EdgeDB server would
+accept password as an environment variable ``EDGEDB_TLS_PRIVATE_KEY_PASSWORD``.
+
+If specified, ``--tls-ciphers`` will be used to call
+``ssl.SSLContext.set_ciphers()``.
```

Comment:
Also, do we need this as a parameter? IMO it makes sense to make this a config entry (that's what Postgres does with `ssl_ciphers` https://www.postgresql.org/docs/10/ssl-tcp.html)